leader-election-impl/src/main/java/com/palantir/leader/AsyncLeadershipObserver.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import com.palantir.common.concurrent.PTExecutors;
+
+public class AsyncLeadershipObserver implements LeadershipObserver {
+
+    private ExecutorService executorService = PTExecutors.newSingleThreadExecutor(true);
+    private List<Runnable> leaderTasks = new ArrayList<>();
+    private List<Runnable> followerTasks = new ArrayList<>();
+
+    @Override
+    public void gainedLeadership() {
+        executorService.execute(this::executeLeaderTasks);
+    }
+
+    @Override
+    public void lostLeadership() {
+        executorService.execute(this::executeFollowerTasks);
+    }
+
+    @Override
+    public synchronized void executeWhenGainedLeadership(Runnable task) {
+        leaderTasks.add(task);
+    }
+
+    @Override
+    public synchronized void executeWhenLostLeadership(Runnable task) {
+        followerTasks.add(task);
+    }
+
+    private void executeFollowerTasks() {
+        followerTasks.forEach(Runnable::run);
+    }
+
+    private void executeLeaderTasks() {
+        leaderTasks.forEach(Runnable::run);
+    }
```

Comment:
I think we should use a SynchronizedList here. The main issue is one of visibility; if in one thread I register a callback and later in another thread I call executeLeaderTasks(), if I'm not wrong it is actually permissible to do nothing; there isn't a happens-before relationship between the read of `leaderTasks` when scheduling them, and the write the callback made. 

Also, if someone registers a callback while we're iterating I think it's possible to get `ConcurrentModificationException` here.

Reply:
Changed the implementation to use synchronized methods, and also utilize a synchronized list. Also to fix the tests, I used a `DeterministicScheduler` and waited until tasks are finished.

Reply:
If I'm not wrong the private methods still need to be synchronized, because the calls to `gainedLeadership` and `lostLeadership` submit things to an executor which might iterate over the list without holding the lock for the list (so you could still get a ConcurrentModificationException; a synchronized list doesn't protect us against that).

Since we don't register too many things it might be better to just use a `CopyOnWriteArrayList` - sorry for bad/ambiguous advice above.